rllib/core/learner/learner.py

```diff
@@ -1147,6 +1148,17 @@ def update(
         # device (e.g. GPU). We move the batch already here to avoid having to move
         # every single minibatch that is created in the `batch_iter` below.
         batch = self._convert_batch_type(batch)
+        for pid, policy_batch in batch.policy_batches.items():
+            if self.module._rl_modules[pid].is_recurrent():
+                assert (
+                    policy_batch.get(SampleBatch.SEQ_LENS, None) is not None
+                ), "Recurrent modules require a `seq_lens` tensor in the policy batch."
+                # We assume that arriving batches for recurrent modules are already
+                # padded to the max sequence length and have tensors of shape
+                # [B, T, ...]. Therefore, we slice sequence lengths in B. See
+                # SampleBatch for more information.
+                policy_batch._slice_seq_lens_in_B = True
```

Comment:
`_slice_seq_lens_in_B` is needed because in the past, SampleBatch assumed that all tensors where 
[B, <feature_dims>], instead of [B, T, <feature_dims>].

Under the old scheme, slicing sequence lengths would work differently.
This is the _only_ place that needs SampleBatch to do this and it's only needed until we get rid of SampleBatch.